src/libutil/source-accessor.hh

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "canon-path.hh"
+#include "hash.hh"
+
+namespace nix {
+
+/**
+ * A read-only filesystem abstraction. This is used by the Nix
+ * evaluator and elsewhere for accessing sources in various
+ * filesystem-like entities (such as the real filesystem, tarballs or
+ * Git repositories).
+ */
+struct SourceAccessor
```

Comment:
Can `FSAccessor` in `src/libstore/fs-accessor.hh` also use this?

Reply:
Unifying those would be a nice followup.

Reply:
I am thinking more than being a nice extra feature, it is a good trial of this abstraction to make sure it is versatile enough.

I think I will take a crack at it and send you a PR against this PR (assuming I get it done in time), and then we have a chance to review everything at once.